### PR TITLE
Simplify lazy semantics

### PR DESCRIPTION
**Pull request by kayceesrk**
Simplify lazy and fix a bug based on discussions from https://github.com/ocaml-multicore/ocaml-multicore/issues/749 and https://github.com/ocaml-multicore/ocaml-multicore/issues/750. 

* Mark lazy as not concurrency (domains, systhreads, fibers) safe.
* Remove RacyLazy exception. Any concurrent use of lazy value may raise Undefined exception.
* Remove domain-local id to distinguish concurrent and recursive forcing.
* Remove try_force.

The commit also fixes a bug to prevent short-circuiting forcing_tag values.

We need to bootstrap to remove references to `RacyLazy` from the bootstrap binaries. After bootstrap, the primitive
`caml_ml_domain_unique_token` should be removed from the runtime.

### CODE REVIEW

**Comment by kayceesrk** [comment]
@gasche would appreciate your review.

File: runtime/domain.c
Comment (gasche) [inline runtime/domain.c:806]: I suppose that the function here is left around to avoid having to do a bootstrap?

File: runtime/obj.c
Comment (gasche) [inline runtime/obj.c]: The logic here is simpler, I don't think the change introduces any bug.

Two questions, unrelated to this PR but to the function itself:
- Do we gain anything by distinguish Forward from any other shortcut tag? I would have expected the first check to be just on `tag != Lazy_tag && tag != Forcing_tag`. (Things are different after the CAS, where we know that the GC has not run since the tag check, so we have fine-grained knowledge that shortcutting hasn't happened which justifies the more precise `CAMLassert(tag == Forcing_tag)`.)
- if the CAS fails, is it really the case that we can assume `tag != Lazy_tag`? I would think that a concurrently-forcing domain could still revert the tag to `Lazy` if the thunk fails with an exception.

File: runtime/sys.c
Comment (gasche) [inline runtime/sys.c]: I'm not sure why these changes are in the PR, maybe some typo-check fix or automated formatting. No problem with having them either.

File: runtime/minor_gc.c
Comment (gadmm) [inline runtime/minor_gc.c:341]: Nice catch! There is similar logic in `runtime/{extern,major_gc,weak}.c` which needs to be fixed in the same way.

File: stdlib/camlinternalLazy.ml
Comment (gadmm) [inline stdlib/camlinternalLazy.ml:83]:
```suggestion
let force_gen ~only_val (lzv : 'arg lazy_t) =
  let lzv = Sys.opaque_identity lzv in
```
The above fix was introduced in trunk to prevent flambda from forwarding the tag information from a previous read (which can become invalid with short-circuiting). (See https://github.com/ocaml/ocaml/issues/7301 and https://github.com/ocaml/ocaml/issues/713.) Do you need something similar? Are there risks that similar issues happen with the new tag Forcing_tag? (Note: this can be addressed in a separate PR if needed.)

**Review by gasche — APPROVED** [review]
Thanks for the ping! (I don't get notifications for ocaml-multicore/)

The change looks correct to me. It does simplify the implementation.

The documentation is changed so that Lazy.force is "unspecified" (but safe) in the case of concurrent forcing. I think we could be more precise, and will try to make some proposals.

The PR comes with some changes in the testsuite, some for backward-compatibility, some (dividing iteration counts by 10 or removing some `cpu_relax()` call) that seem to come from performance considerations, I'm not sure which/why.

**Comment by gasche** [comment]
... after thinking more about it, I think that the current non-documentation of concurrent-forcing behavior is okay.

File: stdlib/lazy.mli
Comment (gadmm) [inline stdlib/lazy.mli]: Beware of line breaks inside `[]` code blocks, they do not display as you want in the documentation.

File: stdlib/camlinternalLazy.ml
Comment (gadmm) [inline stdlib/camlinternalLazy.ml:86]: I think the boundary of the critical section was clearer before: if there is a polling after Obj.tag returns, then there is a bug.
I was looking into expressing this with the `[@poll error]` attribute (to be sure flambda understands this critical section), but it appears unsufficiently expressive. We need one of the extension that I suggested (or Leo's variant of it). The problem is already there in trunk, so I'll see if I have time to open an issue upstream about it.

File: stdlib/camlinternalLazy.ml
Comment (gadmm) [inline stdlib/camlinternalLazy.ml:92]: I am trying to convince myself that this is correct according to OCaml's memory model, but I do not see what ensures that the field is seen as set despite the tag being seen as a forward_tag. Don't you need a load/load barrier before reading the field, or reading the tag with an acquire load, or something stronger? (To my knowledge, OCaml's memory model does not have relaxed atomics yet, so furthermore only a plain atomic read of the tag makes sense here yet.) What makes this correct (and the similar code inlined by lambda)?

**Review by gadmm — COMMENTED** [review]
I have more questions about this code now that I understand it better. Some comments are not about behaviour introduced by this PR, feel free to address them later/separately.

File: stdlib/camlinternalLazy.ml
Comment (gasche) [reply stdlib/camlinternalLazy.ml:92]: My understanding is that the memory model gives a message-passing semantics to non-atomic writes+reads that are synchronized through an atomic at another location.

In the operational semantics of the memory-model paper: if I write to an atomic location, the location's frontier is updated with my frontier, including non-atomic writes that I did before (in program order). Then when the other domain reads from this atomic location, its frontier is updated by my frontier, so it also sees my non-atomic writes to other locations.

File: runtime/obj.c
Comment (gadmm) [reply runtime/obj.c]: I think you are right (the assertion `tag == Forward_tag` can fail), but this is related to this PR.

File: stdlib/camlinternalLazy.ml
Comment (gadmm) [reply stdlib/camlinternalLazy.ml:92]: Sure, but for this you need an atomic read as I wrote

File: stdlib/camlinternalLazy.ml
Comment (kayceesrk) [reply stdlib/camlinternalLazy.ml:83]: I'll merge this suggestion. And I think the rest of the discussion may happen at ocaml/ocaml after the MVP merge to trunk, before the 5.00 release. The MVP will not have flambda support. If the discussion and the PR happens after merge, we will have more eyes on the flambda support as a whole.

File: runtime/domain.c
Comment (kayceesrk) [reply runtime/domain.c:806]: Yes. That's right. Once the changes are ok. I'll do a bootstrap commit and remove this definition.

File: runtime/obj.c
Comment (kayceesrk) [reply runtime/obj.c]: Firstly, my overall idea with this function is to leave some of the functionality intact even though `camlinternalLazy.ml` only cares about `0` and `not 0` cases. The idea is that future implementation of concurrency safe lazy may need to distinguish the various cases. Hence, my choice was to leave the structure of the function as is. 

It is useful to distinguish `Forward` from other shortcut tags; if the result is former and you have the assurance that the GC hasn't occurred, which you currently, the code in `camlinternalLazy.ml` can read the first field of the `Forward` object and return it as a result. 

As @gadmm, the assertion is a bug. 

My assessment is that, rather than think about tricky corner cases, we may simplify the function such that it just returns `0` if the function succeeds in atomically updating the tag from `Lazy` to `Forcing`, and `1` in all the other cases. This way we get to a state where we can merge multicore MVP trunk, and further development can happen on ocaml/ocaml. Thoughts?

File: runtime/sys.c
Comment (kayceesrk) [reply runtime/sys.c]: `check-typo` fixes.

File: stdlib/lazy.mli
Comment (kayceesrk) [reply stdlib/lazy.mli]: Good point. Let me undo it.

File: stdlib/camlinternalLazy.ml
Comment (kayceesrk) [reply stdlib/camlinternalLazy.ml:86]: I'll move it to where it was. `[@poll error]` is not sufficient as it cannot reason about calls to functions IINM. For example, The analysis would encounter the function call `Obj.obj` and report that it is not safe. Is that correct @sadiqj?

File: stdlib/camlinternalLazy.ml
Comment (kayceesrk) [reply stdlib/camlinternalLazy.ml:92]: We should be using C11 memory model to be reasoning about this as the tag update `update_to_forward` and `Obj.tag` are both external calls implemented in C. `update_to_forward` uses `atomic_compare_and_exchange_strong`, but `Obj.tag`, which calls `caml_obj_tag` uses non-atomic reads. My suggestion is to replace the `Tag_val` in `obj.c:obj_tag` with an acquire load. This should be sufficient to ensure that the message-passing semantics works between the tag update and the field write. @stedolan?

**Comment by kayceesrk** [comment]
Also, thanks @gasche and @gadmm for your high-quality reviews.

File: runtime/minor_gc.c
Comment (kayceesrk) [reply runtime/minor_gc.c:341]: Done.

File: runtime/obj.c
Comment (kayceesrk) [reply runtime/obj.c]: Simplified `update_to_forcing` in the latest commits.

File: stdlib/camlinternalLazy.ml
Comment (kayceesrk) [reply stdlib/camlinternalLazy.ml:86]: Moved.

File: stdlib/camlinternalLazy.ml
Comment (kayceesrk) [reply stdlib/camlinternalLazy.ml:92]: Updated the tag read to be load acquire. See https://github.com/ocaml-multicore/ocaml-multicore/pull/760/commits/57137194f2be0caec3697a5980d0851763649d34#diff-5dfaf7e6dc66b87cf561b38c634d235d2e5d747c2a9e27df18f290bc223533cfR43-R44.

Comment (gasche) [reply stdlib/camlinternalLazy.ml:92]: I'm a bit nervous about the performance implications of making `Obj.tag` an atomic read all over the place, "just" for the need of lazy values -- I hope it is uncommon to concurrently change a value tag, and it's not clear that we should pay the price for all tag lookups. Wouldn't it be more prudent to expose an `obj_tag_acquire` variant (or `obj_tag_atomic`) and use it?

File: runtime/obj.c
Comment (gasche) [reply runtime/obj.c]: The simplified version is very nice, thanks!

File: stdlib/camlinternalLazy.ml
Comment (gasche) [reply stdlib/camlinternalLazy.ml:92]: Note: `Obj.tag` is already known as a not-so-efficient way to read tags, but currently there is no faster/simpler primitive (so it remains used in all tag-manipulating OCaml code). And whether or not `Obj.tag` is slow, I think that it may be beneficial to force ourselves to express the intent that the block we are accessing may be concurrently mutated.

Comment (kayceesrk) [reply stdlib/camlinternalLazy.ml:92]: Added `caml_obj_tag_atomic` for this. I've unconditionally used this in lazy implementation, as I'd rather err on the side of safety than performance for lazy for MVP.

File: runtime/obj.c
Comment (kayceesrk) [reply runtime/obj.c]: It needs one extra check to ensure that the value we are attempting to update the tag of is actually a block. I've added this check in `update_to_forcing`.

**Comment by kayceesrk** [comment]
The failures are due to line number differences in the backtrace tests. I've already fixed these twice in this PR :-| I'll wait until the review on the changes reaches a fixed point, and then I shall fix the failing tests.

File: stdlib/camlinternalLazy.ml
Comment (gadmm) [reply stdlib/camlinternalLazy.ml:83]: Let's remind ourselves to keep the original comment describing the problem when merging upstream, and to go back to it later.

File: stdlib/camlinternalLazy.ml
Comment (gadmm) [reply stdlib/camlinternalLazy.ml:86]: Trying on trunk (similar code), the poll-error attribute does not see Obj.tag or Obj.obj as polling, but it does see Obj.field as polling (whether it actually polls or not on trunk, there is something to do about it).

You still need a notation to allow a call to a polling function as suggested here: https://github.com/ocaml/ocaml/pull/10462#issuecomment-868457097 (namely the call to `force_gen_lazy_block`) because there is no obvious way to refactor it (apart maybe by conveying the need to evaluate the thunk using an exception?).

File: stdlib/camlinternalLazy.ml
Comment (gadmm) [reply stdlib/camlinternalLazy.ml:92]: Don't forget that there is similar code in `lambda/matching.ml` which will need an acquire load as well.

Alternatively, you only need a load/load barrier (e.g. `dmb ishld` on aarch64) before reading the field, which removes the synchronisation cost for short-circuited blocks (though it is likely very small).

Eventually it would be nice to have acquire/release atomics in the OCaml memory model, or barriers to reason about non-atomic locations like here, because it looks complicated to reason about mixing two memory models (and especially that this does not introduce any catch-fire semantics).

Note that since the header is used as an atomic location by the runtime, the tag must always be read from C with at least a `relaxed` atomic read (or else it can catch fire).

Comment (gadmm) [reply stdlib/camlinternalLazy.ml:92]: To clarify, my comment about mixing memory models concerns the inlined forcing code at `matching.ml`.

Acquire/release fences are the most natural in my alternative solution at https://gitlab.com/gadmm/stdlib-experiment/-/blob/master/other/sync_lazy.ml (which also avoids mutating the tag concurrently); and an OCaml version of a release fence might only need a `dmb st` fence.

Comment (kayceesrk) [reply stdlib/camlinternalLazy.ml:92]: I've already put in the inlined fixes for `lambda/matching.ml`. See https://github.com/ocaml-multicore/ocaml-multicore/pull/760/files#diff-059d6544bdc63e5f278818ea9f50bef7aa3f855c4d5ce6fd726abee18b187f55R1860-R1861. Am I missing something?

Comment (kayceesrk) [reply stdlib/camlinternalLazy.ml:92]: I'm not comfortable using raw fences in the OCaml. Our PLDI'18 paper does not include them, and we'd have to extend the work to talk about the guarantees, and in particular, preserving LDRF and their interaction with the rest of the memory model. I wouldn't think of adding fences lightly. 

I would like to aim for a working implementation in this PR, even if it is not the most efficient, and save the optimisations for after multicore MVP is merged with OCaml.

Comment (gadmm) [reply stdlib/camlinternalLazy.ml:92]: To be more precise, `inline_lazy_force_switch` in `matching.ml` does not use `prim_obj_tag` so I assume that it makes an unsynchronized read. This is the one used to compile native code (for efficiency). Possible solutions:
1. replace it with a call to `inline_lazy_force_cond` (remove optimisation)
2. add an argument to Lswitch to read-acquire the tag (heavy solution with a single use-case)
3. add a load-load fence before reading the field (only minimal working solution of a fence for internal use needed in a first time, more carefully design can come later if programmers need it)
4. use my variant above that stores the forward bit in the thunk field rather than with the tag, and atomically read this field.

If you do not want a fence, then 1. and 4. are the simplest I see (and we could do 1. first and experiment with 4. later).

Comment (gadmm) [reply stdlib/camlinternalLazy.ml:92]: I understand the “future work” paragraph p.14 of your paper to refer to adding acquire/release primitives in the future, where the mentioned Kang et al. paper considers fences (IIANM), so I was exploring this possibility. But whatever suitable acquire/release primitives would be better than SC atomics in the current use-case. (Any idea @stedolan?)

Anyway my idea was not to suggest that doing this work is any urgent, but that for now it would be useful to have something equivalent to what you can already do in the C runtime but without incurring a C call, reserved for internal use. I think that the LDRF property is trivially preserved if you see acquire/release atomics as non-atomic, and fences as noops (of course refining the LDRF property and proving that the extended memory model has nice properties is substantially more work).

**Comment by gasche** [comment]
My high-level impression is that we should try to merge a "good first part" version of this PR, that simplifies the current implementation and does not introduce any new bugs, and then have a separate, second discussion about how to safely read concurrently-modified tags. (I suspect we may not be out of surprises there: many other parts of the runtime read tags, how safe are they with respect to concurrent mutation?)

File: stdlib/camlinternalLazy.ml
Comment (xavierleroy) [reply stdlib/camlinternalLazy.ml:92]: > Added caml_obj_tag_atomic for this.

Apologies for yet another dissenting opinion, but I don't like the idea of having multiple versions of C functions that differ only in the memory model used.

First, it's a software engineering nightmare: code duplication + tiny changes...  We already have enough smelly C code to maintain, thank you very much for not adding more.

Second, the performance degradation that @gasche is worried about has not been demonstrated.  It will be null on x86 (all loads have acquire behaviors) and (in my guess) negligible on other platforms, since `caml_obj_tag` is quite expensive already.

Third, I think the C primitives should be made to fit the Multicore OCaml memory model, and not the other way around.  Meaning that if a C11 nonatomic load in a C primitive is a source of concern, it makes perfect sense to make this load more memory coherent for all uses.  And perhaps other primitives from Obj should be scrutinized from this angle too.

Comment (gasche) [reply stdlib/camlinternalLazy.ml:92]: Here is a proposal for `inline_lazy_force_switch`:

```diff
diff --git a/lambda/matching.ml b/lambda/matching.ml
index 985f86e25..d996489d2 100644
--- a/lambda/matching.ml
+++ b/lambda/matching.ml
@@ -1943,12 +1943,12 @@ let inline_lazy_force_switch arg loc =
         ( Lprim (Pisint, [ varg ], loc),
           varg,
           Lswitch
-            ( varg,
-              { sw_numconsts = 0;
-                sw_consts = [];
-                sw_numblocks = 256;
+            ( Lprim (Pccall prim_obj_tag, [ varg ], loc),
+              { sw_numblocks = 0;
+                sw_blocks = [];
+                sw_numconsts = 256;
                 (* PR#6033 - tag ranges from 0 to 255 *)
-                sw_blocks =
+                sw_consts =
                   [ (Obj.forward_tag, Lprim (Pfield(0, Pointer, Mutable),
                                              [ varg ], loc));
```

Unfortunately, the compiler generates not-so-good code in this case (it handles poorly a sparse switch with only integer constants). I have some matching-specific ideas to improve it, but it should be discussed with matching experts and is somewhat irrelevant to the current PR discussion. I believe that this version is better than reverting to `inline_lazy_force_cond`, so it would be a good first step for that function.

**Comment by gasche** [comment]
To continue on my previous line of thought: I think we could cut the current PR to the  commit before "Add new primitive to atomically read tag", and get this merged right away, and then have a dedicated discussion on the memory-safety to make sure that we properly address the issue.

(@kayceesrk : just checking that you know about `make promote DIR=...` to auto-fix line numbers and in generate test outputs in the testsuite.)

**Comment by gadmm** [comment]
I think we're there right after changing `lazy_..._switch` to `lazy_..._cond` or to what @gasche proposed, and changing obj_tag to address @xavierleroy's concern. It's an easy fix so no need to delay the last part of this PR.

File: stdlib/camlinternalLazy.ml
Comment (gasche) [reply stdlib/camlinternalLazy.ml:92]: @xavierleroy I'm nervous because `Obj.tag`-like functionality is used for pattern-matching, so actually quite heavily used in OCaml programs. Something smells bad about the idea that the dirty tag tricks we use for lazy thunks spreads overhead into all OCaml programs, even those that never use Lazy. (Maybe the cost is neglectible but, if we are going to pay it, we should measure it!)

More importantly, we should think hard about whether we can actually safely mutate tags in this way for lazy thunks, and which parts of the runtime are affected. Adding a barrier to `Obj.tag` is not going to make the problem go away. Having two different primitives, and forcing us to think about which one to use where, is a way to record our decisions/understanding about this problem. Maybe in the end we would find out that we only use the atomic version, and we can simplify again, but in the meantime we should have a clear way to tell which parts of the codebase have to account for concurrent tag modifications, and which do not.

Comment (xavierleroy) [reply stdlib/camlinternalLazy.ml:92]: > I'm nervous because Obj.tag-like functionality is used for pattern-matching

Can you be more precise?  Does the pattern-matching compiler generate calls to `caml_obj_tag` or similar C primitives?  The "get tag" operations produced to implement Lswitch, etc, are under the control of the OCaml compiler, so they obey the Multicore OCaml memory model.

Comment (gasche) [reply stdlib/camlinternalLazy.ml:92]: I reviewed the use of `Obj.tag` in the compiler codebase; @xavierleroy is correct (of course!) that there is mostly nothing performance-sensitive there, and that adding an atomic load to all `Obj.tag` users would be fine. The use-cases of `caml_obj_tag`  in `lambda/matching.ml` come from the compilation of `lazy` patterns, and I guess it's okay if those suffer from overhead. The compilation of non-lazy patterns uses a different `get_tag` operation.

> The "get tag" operations produced to implement Lswitch, etc, are under the control of the OCaml compiler, so they obey the Multicore OCaml memory model.

Right but, if I understand correctly, as @gadmm pointed out the Multicore OCaml memory model says little about what you get if you non-atomically read a field (as `get_tag` does) that was atomically-written to by a Lazy mutation (as `Lazy.force` does). My understanding is that calling `get_tag` to learn that a block has one of the lazy tags (Lazy, Forcing, Forward) *would* be racy and not guarantee anything about the block arguments. We are safe because we never do this -- and my unboxed-constructors work is also safe, as `Lazy.t` is known to have any shape due to Forward shortcutting.

**Comment by gasche** [comment]
What about the lazy-tag checks in other parts of the runtime?

- compare.c, hash.c: I think they need an atomic read to handle Forward_tag safely
- extern.c: this logic can run concurrently with another mutator, but iirc. we decided that it is unspecified what happens in this case (but no crashes); I think the current code will not crash if it reads the tags wrong, at worst it will incorrectly shortcut a pointer in the serialized result. No change seems required.
- finalise.c: no change is required
- minor_gc.c, major_gc.c, weak.c: no change required as iirc. no mutators can run concurrently with the GC

**Comment by bluddy** [comment]
> compare.c, hash.c: I think they need an atomic read to handle Forward_tag safely

Why? Worst case you get the wrong result (which is fine), but do you get a crash?

**Comment by gasche** [comment]
I think you are correct: we can get garbage in this case, but no crash, because the block arguments of a lazy value remains well-formed at any point. (With the new policy, specified in the current PR, that lazy values are not thread-safe, this is fine. Previously we tried to make them shareable, as long as only there is no racy force.)

**Comment by kayceesrk** [comment]
Based on the recommendations, I have 

1. Removed the new primitive to atomically read tag and made `obj_tag` load the tag using acquire.
2. Incorporated @gasche's suggestion for `inline_lazy_force_switch`

>  just checking that you know about make promote DIR=... to auto-fix line numbers and in generate test outputs in the testsuite.

I did not know this. Thanks for pointing this out.

File: stdlib/camlinternalLazy.ml
Comment (gadmm) [reply stdlib/camlinternalLazy.ml:92]: It is true that the memory model does not say what happens when treating an atomic location as non-atomic (maybe the model can trivially be extended to say what happens in this case), but I was more pointing out that would be racy anyway (and the alternative of a strong atomic load is ~not nice~ totally worth exploring).

File: runtime/obj.c
Comment (gadmm) [inline runtime/obj.c:46]:
```suggestion
    /* The acquire load ensures that reading the field of a Forward_tag
       block has appropriate synchronization. */
    hd = (header_t)atomic_load_acq(Hp_atomic_val(arg));
```

**Review by gadmm — APPROVED** [review]
LGTM. I would just add a comment (below).

**Comment by gasche** [comment]
(Note: my inline_lazy_switch suggestion requires massaging from the matching+backend people; if we merge as-is, as I think is reasonable, I will open an issue to track this.)

**Comment by gadmm** [comment]
> (Note: my inline_lazy_switch suggestion requires massaging from the matching+backend people; if we merge as-is, as I think is reasonable, I will open an issue to track this.)

Or that could be used as an incentive to explore an implementation that does not mutate tags!

File: runtime/obj.c
Comment (kayceesrk) [reply runtime/obj.c:46]: Good point. Added comment in https://github.com/ocaml-multicore/ocaml-multicore/pull/760/commits/c45feb0653bf56adac0ccd5568b035b87b13d47b that also says where the field read occurs.

**Comment by kayceesrk** [comment]
Will merge when the CI turns green. Thanks @gasche, @gadmm and @xavierleroy.

**Pull request merged**
